pkg/controller/master/upgradelog/controller.go

```diff
@@ -0,0 +1,625 @@
+	upgradeLogClusterFlowReady        = "ClusterFlowReady"
+	upgradeLogClusterOutputReady      = "ClusterOutputReady"
+
+	// Annotation indicating the state of UpgradeLog
+	upgradeLogStateAnnotation = "harvesterhci.io/upgradeLogState"
+	upgradeLogStateCollecting = "Collecting"
+	upgradeLogStateStopped    = "Stopped"
+
+	// Logging infra images
+	fluentBitImageRepo      = "rancher/mirrored-fluent-fluent-bit"
+	fluentBitImageTag       = "1.9.5"
+	fluentdImageRepo        = "rancher/mirrored-banzaicloud-fluentd"
+	fluentdImageTag         = "v1.14.6-alpine-5"
+	configReloaderImageRepo = "rancher/mirrored-jimmidyson-configmap-reload"
+	configReloaderImageTag  = "v0.4.0"
+)
+
+type handler struct {
+	ctx                 context.Context
+	namespace           string
+	addonCache          ctlharvesterv1.AddonCache
+	clusterFlowClient   ctlloggingv1.ClusterFlowClient
+	clusterOutputClient ctlloggingv1.ClusterOutputClient
+	daemonSetClient     ctlappsv1.DaemonSetClient
+	daemonSetCache      ctlappsv1.DaemonSetCache
+	deploymentClient    ctlappsv1.DeploymentClient
+	jobClient           ctlbatchv1.JobClient
+	jobCache            ctlbatchv1.JobCache
+	loggingClient       ctlloggingv1.LoggingClient
+	managedChartClient  ctlmgmtv3.ManagedChartClient
+	managedChartCache   ctlmgmtv3.ManagedChartCache
+	pvcClient           ctlcorev1.PersistentVolumeClaimClient
+	serviceClient       ctlcorev1.ServiceClient
+	statefulSetClient   ctlappsv1.StatefulSetClient
+	statefulSetCache    ctlappsv1.StatefulSetCache
+	upgradeClient       ctlharvesterv1.UpgradeClient
+	upgradeCache        ctlharvesterv1.UpgradeCache
+	upgradeLogClient    ctlharvesterv1.UpgradeLogClient
+	upgradeLogCache     ctlharvesterv1.UpgradeLogCache
+}
+
+func (h *handler) OnUpgradeLogChange(_ string, upgradeLog *harvesterv1.UpgradeLog) (*harvesterv1.UpgradeLog, error) {
+	if upgradeLog == nil || upgradeLog.DeletionTimestamp != nil {
+		return upgradeLog, nil
+	}
+	logrus.Infof("Processing UpgradeLog %s/%s", upgradeLog.Namespace, upgradeLog.Name)
+
+	// Initialize the UpgradeLog resource
+	if harvesterv1.UpgradeLogReady.GetStatus(upgradeLog) == "" {
+		logrus.Info("Initialize UpgradeLog")
+		toUpdate := upgradeLog.DeepCopy()
+		harvesterv1.UpgradeLogReady.CreateUnknownIfNotExists(toUpdate)
+		return h.upgradeLogClient.Update(toUpdate)
+	}
+
+	// Try to bring up the logging operator by installing rancher-logging ManagedChart
+	if harvesterv1.LoggingOperatorDeployed.GetStatus(upgradeLog) == "" {
+		logrus.Info("Check if there are any existing logging-operator")
+
+		toUpdate := upgradeLog.DeepCopy()
+		harvesterv1.LoggingOperatorDeployed.CreateUnknownIfNotExists(toUpdate)
+
+		// Detect rancher-logging Addon
+
+		addon, err := h.addonCache.Get(util.CattleLoggingSystemNamespaceName, util.RancherLoggingName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+			logrus.Info("rancher-logging Addon is not installed")
+		} else {
+			if addon.Status.Status == harvesterv1.AddonEnabled {
```

Comment:
Checking the current state instead of the desired state is preferred. I'm not sure if there is any chance the `addon.Spec.Enabled` is `true` but the `addon.Status.Status` is `false`. If that won't happen, we can go with `addon.Spec.Enabled` without problems.

Or do you suggest, in the first place, we should trust the `rancher-logging` addon will eventually roll out by its controller and will become active in `addon.Status.Status` in case we mistakenly deploy a duplicated logging-operator?

Reply:
Spec is from user input, it is the desired result, the controller is reconcilling to align with the spec, status reflect a snapshot in the timeline, it could be that when you check it, it is not ready, but soon it becomes ready

here we could blindly believe the addon controller will roll the addon out, otherwise, its addon's bug

Reply:
Make sense to me. I updated it, thanks.